src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -675,80 +649,67 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
         case app @ Apply(fun, args) =>
           val sym = fun.symbol
 
-          if (sym.isLabel) {  // jump to a label
+          if (sym.isLabel) { // jump to a label
             genLoadLabelArguments(args, labelDef(sym), app.pos)
             bc goTo programPoint(sym)
           } else if (isPrimitive(sym)) { // primitive method call
             generatedType = genPrimitiveOp(app, expectedType)
-          } else {  // normal method call
-
-            def genNormalMethodCall() {
-
-              val invokeStyle =
-                if (sym.isStaticMember) InvokeStyle.Static
-                else if (sym.isPrivate || sym.isClassConstructor) InvokeStyle.Special
-                else InvokeStyle.Virtual
-
-              if (invokeStyle.hasInstance) {
-                genLoadQualifier(fun)
+          } else { // normal method call
+            val invokeStyle =
+              if (sym.isStaticMember) InvokeStyle.Static
+              else if (sym.isPrivate || sym.isClassConstructor) InvokeStyle.Special
+              else InvokeStyle.Virtual
+
+            if (invokeStyle.hasInstance) genLoadQualifier(fun)
+            genLoadArguments(args, paramTKs(app))
+
+            val Select(qual, _) = fun // fun is a Select, also checked in genLoadQualifier
+            if (sym == definitions.Array_clone) {
+              // Special-case Array.clone, introduced in 36ef60e. The goal is to generate this
+              // call as "[I.clone" instead of "java/lang/Object.clone" (which would also work).
+              // Not sure why this is useful, maybe for performance. The commit message says
```

Comment:
Checked again the JVM spec

[Access Control](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-5.html#jvms-5.4.4): protected method R, declared in C, is accessible to a class or interface D if
  - D is a subclass of C, and
  - R must contain a symbolic reference to a class T, such that T is either a subclass of D, a superclass of D, or D itself

"symbolic reference to class T" means the receiver class in the invocation instruction.

But there's more: "This discussion omits a related restriction [...] is checked as part of the verification process (§4.10.1.8); it is not part of link-time access control."

[4.10.1.8. Type Checking for protected Members](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.10.1.8) encodes that a protected method C.m can be accessed in a subclass D only if the receiver object's type conforms to D. This corresponds to the [Java spec 6.6.2.1](https://docs.oracle.com/javase/specs/jls/se8/html/jls-6.html#jls-6.6.2.1).

Assume we have
```scala
class C { override def clone(): Object = "hi" }
```

We cannot emit `def f(c: C) = c.clone()` as `Object.clone()`. he JVM would perform the following checks:
  - [Method Resolution (5.4.3.3)](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-5.html#jvms-5.4.3.3) finds `Object.clone`
  - [Access Control (5.4.4)](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-5.html#jvms-5.4.4) is checked at the end of resolution and succeeds (every class and interface is a subclass of `Object`, so `Object.clone` can always passes)
  - [Verification (4.10.1.8)](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.10.1.8) will fail the `passesProtectedCheck`

I tested this with a compiler hack, we get `java.lang.VerifyError` indeed.

So the reason `Object.clone` works for arrays is really the special cases you point to above. Given that javac also emits `[I.clone` we better stick to that.

Finally, to make the discussion complete, here's the example with traits:

```
trait T { override def clone(): Object = "hu" }
trait U extends T
class K extends U
object T1 { def f1(u: U) = u.clone() }
```

We have to emit `u.clone` as `T.clone`. Using `U.clone` gives `NoSuchMethodError: U.clone`. Looking at [5.4.3.4. Interface Method Resolution](https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-5.html#jvms-5.4.3.4) I actually don't see why this is the case, it might be a bug in the JVM.
  -